lib/models/chat_history.dart

```diff
@@ -222,6 +224,70 @@ Stream<DeltaEvent> _handleDocumentChange(
           messageId: change.doc.id,
           pinned: false);
       yield AppendDeltaEvent(model);
+      break;
+    case "channel.hype_train.begin":
+      final model = TwitchHypeTrainEventModel(
+          pinned: true,
+          messageId: "train${data['event']['id']}",
+          level: 1,
+          progress: data['event']['progress'],
+          goal: data['event']['goal'],
+          total: data['event']['total']);
+      yield AppendDeltaEvent(model);
+      break;
+    case "channel.hype_train.progress":
+      yield UpdateDeltaEvent("train${data['event']['id']}", (message) {
+        if (message is! TwitchHypeTrainEventModel) {
+          return message;
+        }
+        // Since you can receive duplicates and order is not guaranteed
+        final level = data['event']['level'];
```

Comment:
Also worth moving this logic to the model, so we can do something like:

```
if (message is! TwitchHypeTrainEventModel) {
  return message;
}
return message.withProgress(change.doc);
```

where `withProgress` performs this logic and returns a new model if necessary.